// @yaga/leaflet-ng2/lib/attribution-control.directive.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var leaflet_1 = require("leaflet");
var consts_1 = require("./consts");
var map_component_1 = require("./map.component");
var AttributionControlDirective = (function (_super) {
    __extends(AttributionControlDirective, _super);
    function AttributionControlDirective(mapComponent) {
        var _this = _super.call(this, { prefix: consts_1.ATTRIBUTION_PREFIX }) || this;
        _this.displayChange = new core_1.EventEmitter();
        _this.zIndexChange = new core_1.EventEmitter();
        _this.positionChange = new core_1.EventEmitter();
        _this.prefixChange = new core_1.EventEmitter();
        _this.attributionsChange = new core_1.EventEmitter();
        _this.addEvent = new core_1.EventEmitter();
        _this.removeEvent = new core_1.EventEmitter();
        _this.clickEvent = new core_1.EventEmitter();
        _this.dbclickEvent = new core_1.EventEmitter();
        _this.mousedownEvent = new core_1.EventEmitter();
        _this.mouseoverEvent = new core_1.EventEmitter();
        _this.mouseoutEvent = new core_1.EventEmitter();
        mapComponent.addControl(_this);
        var self = _this;
        /* tslint:disable:only-arrow-functions */
        _this.onRemove = function () {
            self.displayChange.emit(false);
            self.removeEvent.emit({ target: self, type: 'remove' });
            return self;
        };
        _this.onAdd = function () {
            self.displayChange.emit(true);
            self.addEvent.emit({ target: self, type: 'add' });
            return self.getContainer();
        };
        /* tslint:enable */
        // Events
        _this.getContainer().addEventListener('click', function (event) {
            _this.clickEvent.emit(event);
        });
        _this.getContainer().addEventListener('dbclick', function (event) {
            _this.dbclickEvent.emit(event);
        });
        _this.getContainer().addEventListener('mousedown', function (event) {
            _this.mousedownEvent.emit(event);
        });
        _this.getContainer().addEventListener('mouseover', function (event) {
            _this.mouseoverEvent.emit(event);
        });
        _this.getContainer().addEventListener('mouseout', function (event) {
            _this.mouseoutEvent.emit(event);
        });
        return _this;
    }
    AttributionControlDirective.prototype.ngOnDestroy = function () {
        this._map.removeControl(this);
    };
    AttributionControlDirective.prototype.setPosition = function (val) {
        _super.prototype.setPosition.call(this, val);
        this.positionChange.emit(val);
        return this;
    };
    Object.defineProperty(AttributionControlDirective.prototype, "opacity", {
        get: function () {
            return parseFloat(this.getContainer().style.opacity);
        },
        set: function (val) {
            this.getContainer().style.opacity = val.toString();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AttributionControlDirective.prototype, "display", {
        get: function () {
            return this._map && this.getContainer().style.display !== 'none';
        },
        set: function (val) {
            if (!this._map) {
                // No map available...
                return;
            }
            if (val) {
                this.getContainer().style.display = '';
                return;
            }
            this.getContainer().style.display = 'none';
            return;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AttributionControlDirective.prototype, "position", {
        get: function () {
            return this.getPosition();
        },
        set: function (val) {
            this.setPosition(val);
        },
        enumerable: true,
        configurable: true
    });
    AttributionControlDirective.prototype.setPrefix = function (prefix) {
        _super.prototype.setPrefix.call(this, prefix);
        this.prefixChange.emit(prefix);
        return this;
    };
    Object.defineProperty(AttributionControlDirective.prototype, "prefix", {
        get: function () {
            return this.options.prefix;
        },
        set: function (val) {
            this.setPrefix(val);
        },
        enumerable: true,
        configurable: true
    });
    AttributionControlDirective.prototype.addAttribution = function (val) {
        _super.prototype.addAttribution.call(this, val);
        this.attributionsChange.emit(this.attributions);
        return this;
    };
    AttributionControlDirective.prototype.removeAttribution = function (val) {
        _super.prototype.removeAttribution.call(this, val);
        this.attributionsChange.emit(this.attributions);
        return this;
    };
    Object.defineProperty(AttributionControlDirective.prototype, "attributions", {
        get: function () {
            var keys = Object.keys(this._attributions);
            var arr = [];
            for (var i = 0; i < keys.length; i += 1) {
                if (this._attributions[keys[i]] === 1) {
                    arr.push(keys[i]);
                }
            }
            return arr;
        },
        set: function (val) {
            this.removeAllAttributions(true);
            for (var i = 0; i < val.length; i += 1) {
                _super.prototype.addAttribution.call(this, val[i]);
            }
            this.attributionsChange.emit(this.attributions);
        },
        enumerable: true,
        configurable: true
    });
    AttributionControlDirective.prototype.removeAllAttributions = function (silent) {
        var keys = Object.keys(this._attributions);
        for (var i = 0; i < keys.length; i += 1) {
            _super.prototype.removeAttribution.call(this, keys[i]);
        }
        if (silent) {
            return this;
        }
        this.attributionsChange.emit([]);
        return this;
    };
    return AttributionControlDirective;
}(leaflet_1.Control.Attribution));
AttributionControlDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: 'yaga-attribution-control',
            },] },
];
/** @nocollapse */
AttributionControlDirective.ctorParameters = function () { return [
    { type: map_component_1.MapComponent, decorators: [{ type: core_1.Inject, args: [core_1.forwardRef(function () { return map_component_1.MapComponent; }),] },] },
]; };
AttributionControlDirective.propDecorators = {
    'displayChange': [{ type: core_1.Output },],
    'zIndexChange': [{ type: core_1.Output },],
    'positionChange': [{ type: core_1.Output },],
    'prefixChange': [{ type: core_1.Output },],
    'attributionsChange': [{ type: core_1.Output },],
    'addEvent': [{ type: core_1.Output, args: ['add',] },],
    'removeEvent': [{ type: core_1.Output, args: ['remove',] },],
    'clickEvent': [{ type: core_1.Output, args: ['click',] },],
    'dbclickEvent': [{ type: core_1.Output, args: ['dbclick',] },],
    'mousedownEvent': [{ type: core_1.Output, args: ['mousedown',] },],
    'mouseoverEvent': [{ type: core_1.Output, args: ['mouseover',] },],
    'mouseoutEvent': [{ type: core_1.Output, args: ['mouseout',] },],
    'opacity': [{ type: core_1.Input },],
    'display': [{ type: core_1.Input },],
    'position': [{ type: core_1.Input },],
    'prefix': [{ type: core_1.Input },],
    'attributions': [{ type: core_1.Input },],
};
exports.AttributionControlDirective = AttributionControlDirective;
