// @yaga/leaflet-ng2/lib/circle-marker.directive.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var leaflet_1 = require("leaflet");
var lng2lat_1 = require("./lng2lat");
var map_component_1 = require("./map.component");
// Content-Child imports
var popup_directive_1 = require("./popup.directive");
var tooltip_directive_1 = require("./tooltip.directive");
var CircleMarkerDirective = (function (_super) {
    __extends(CircleMarkerDirective, _super);
    function CircleMarkerDirective(mapComponent) {
        var _this = _super.call(this, [0, 0]) || this;
        _this.displayChange = new core_1.EventEmitter();
        _this.strokeChange = new core_1.EventEmitter();
        _this.colorChange = new core_1.EventEmitter();
        _this.weightChange = new core_1.EventEmitter();
        _this.opacityChange = new core_1.EventEmitter();
        _this.lineCapChange = new core_1.EventEmitter();
        _this.lineJoinChange = new core_1.EventEmitter();
        _this.dashArrayChange = new core_1.EventEmitter();
        _this.dashOffsetChange = new core_1.EventEmitter();
        _this.fillChange = new core_1.EventEmitter();
        _this.fillColorChange = new core_1.EventEmitter();
        _this.fillOpacityChange = new core_1.EventEmitter();
        _this.fillRuleChange = new core_1.EventEmitter();
        // @Output() public rendererChange: EventEmitter<number> = new EventEmitter();
        _this.classNameChange = new core_1.EventEmitter();
        _this.styleChange = new core_1.EventEmitter();
        _this.positionChange = new core_1.EventEmitter();
        _this.latChange = new core_1.EventEmitter();
        _this.lngChange = new core_1.EventEmitter();
        _this.radiusChange = new core_1.EventEmitter();
        /* tslint:disable:max-line-length */
        _this.geoJSONChange = new core_1.EventEmitter();
        /* tslint:enable */
        _this.addEvent = new core_1.EventEmitter();
        _this.removeEvent = new core_1.EventEmitter();
        _this.popupopenEvent = new core_1.EventEmitter();
        _this.popupcloseEvent = new core_1.EventEmitter();
        _this.tooltipopenEvent = new core_1.EventEmitter();
        _this.tooltipcloseEvent = new core_1.EventEmitter();
        _this.clickEvent = new core_1.EventEmitter();
        _this.dbclickEvent = new core_1.EventEmitter();
        _this.mousedownEvent = new core_1.EventEmitter();
        _this.mouseoverEvent = new core_1.EventEmitter();
        _this.mouseoutEvent = new core_1.EventEmitter();
        _this.contextmenuEvent = new core_1.EventEmitter();
        _this.initialized = false;
        _this.feature = _this.feature || { type: 'Feature', properties: {}, geometry: { type: 'Point', coordinates: [] } };
        _this.feature.properties = _this.feature.properties || {};
        _this.on('remove', function () {
            _this.displayChange.emit(false);
        });
        _this.on('add', function () {
            _this.displayChange.emit(true);
        });
        mapComponent.addLayer(_this);
        // Events
        _this.on('add', function (event) {
            _this.addEvent.emit(event);
        });
        _this.on('remove', function (event) {
            _this.removeEvent.emit(event);
        });
        _this.on('popupopen', function (event) {
            _this.popupopenEvent.emit(event);
        });
        _this.on('popupclose', function (event) {
            _this.popupcloseEvent.emit(event);
        });
        _this.on('tooltipopen', function (event) {
            _this.tooltipopenEvent.emit(event);
        });
        _this.on('tooltipclose', function (event) {
            _this.tooltipcloseEvent.emit(event);
        });
        _this.on('click', function (event) {
            _this.clickEvent.emit(event);
        });
        _this.on('dbclick', function (event) {
            _this.dbclickEvent.emit(event);
        });
        _this.on('mousedown', function (event) {
            _this.mousedownEvent.emit(event);
        });
        _this.on('mouseover', function (event) {
            _this.mouseoverEvent.emit(event);
        });
        _this.on('mouseout', function (event) {
            _this.mouseoutEvent.emit(event);
        });
        _this.on('contextmenu', function (event) {
            _this.contextmenuEvent.emit(event);
        });
        return _this;
    }
    CircleMarkerDirective.prototype.ngAfterViewInit = function () {
        this.initialized = true;
        if (this.popupDirective) {
            this.bindPopup(this.popupDirective);
        }
        if (this.tooltipDirective) {
            this.bindTooltip(this.tooltipDirective);
        }
    };
    CircleMarkerDirective.prototype.ngOnDestroy = function () {
        this.removeFrom(this._map);
    };
    CircleMarkerDirective.prototype.setLatLng = function (val) {
        _super.prototype.setLatLng.call(this, val);
        if (!this.initialized) {
            return this;
        }
        this.positionChange.emit(this._latlng);
        this.latChange.emit(this._latlng.lat);
        this.lngChange.emit(this._latlng.lng);
        this.geoJSONChange.emit(this.geoJSON);
        return this;
    };
    Object.defineProperty(CircleMarkerDirective.prototype, "position", {
        get: function () {
            return this._latlng;
        },
        set: function (val) {
            this.setLatLng(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "lat", {
        get: function () {
            return this._latlng.lat;
        },
        set: function (val) {
            this.setLatLng([val, this.lng]);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "lng", {
        get: function () {
            return this._latlng.lng;
        },
        set: function (val) {
            this.setLatLng([this.lat, val]);
        },
        enumerable: true,
        configurable: true
    });
    CircleMarkerDirective.prototype.setRadius = function (val) {
        _super.prototype.setRadius.call(this, val);
        this.radiusChange.emit(val);
        return this;
    };
    Object.defineProperty(CircleMarkerDirective.prototype, "radius", {
        get: function () {
            return this.getRadius();
        },
        set: function (val) {
            this.setRadius(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "geoJSON", {
        get: function () {
            return this.toGeoJSON();
        },
        set: function (val) {
            this.feature.properties = val.properties;
            var geomType = val.geometry.type; // Normally 'Point'
            /* istanbul ignore if */
            if (geomType !== 'Point') {
                throw new Error('Unsupported geometry type: ' + geomType);
            }
            this.setLatLng(lng2lat_1.lng2lat(val.geometry.coordinates));
        },
        enumerable: true,
        configurable: true
    });
    CircleMarkerDirective.prototype.setStyle = function (style) {
        _super.prototype.setStyle.call(this, style);
        if (style.hasOwnProperty('stroke')) {
            this.strokeChange.emit(style.stroke);
        }
        if (style.hasOwnProperty('color')) {
            this.colorChange.emit(style.color);
        }
        if (style.hasOwnProperty('weight')) {
            this.weightChange.emit(style.weight);
        }
        if (style.hasOwnProperty('opacity')) {
            this.opacityChange.emit(style.opacity);
        }
        if (style.hasOwnProperty('lineCap')) {
            this.lineCapChange.emit(style.lineCap);
        }
        if (style.hasOwnProperty('lineJoin')) {
            this.lineJoinChange.emit(style.lineJoin);
        }
        if (style.hasOwnProperty('dashArray')) {
            this.dashArrayChange.emit(style.dashArray);
        }
        if (style.hasOwnProperty('dashOffset')) {
            this.dashOffsetChange.emit(style.dashOffset);
        }
        if (style.hasOwnProperty('fill')) {
            this.fillChange.emit(style.fill);
        }
        if (style.hasOwnProperty('fillColor')) {
            this.fillColorChange.emit(style.fillColor);
        }
        if (style.hasOwnProperty('fillOpacity')) {
            this.fillOpacityChange.emit(style.fillOpacity);
        }
        if (style.hasOwnProperty('fillRule')) {
            this.fillRuleChange.emit(style.fillRule);
        }
        if (style.hasOwnProperty('className')) {
            this.classNameChange.emit(style.className);
        }
        this.styleChange.emit(style);
        return this;
    };
    Object.defineProperty(CircleMarkerDirective.prototype, "opacity", {
        get: function () {
            return this.options.opacity;
        },
        set: function (val) {
            this.setStyle({ opacity: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "stroke", {
        get: function () {
            return this.options.stroke;
        },
        set: function (val) {
            this.setStyle({ stroke: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "color", {
        get: function () {
            return this.options.color;
        },
        set: function (val) {
            this.setStyle({ color: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "weight", {
        get: function () {
            return this.options.weight;
        },
        set: function (val) {
            this.setStyle({ weight: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "lineCap", {
        get: function () {
            return this.options.lineCap;
        },
        set: function (val) {
            this.setStyle({ lineCap: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "lineJoin", {
        get: function () {
            return this.options.lineJoin;
        },
        set: function (val) {
            this.setStyle({ lineJoin: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "dashArray", {
        get: function () {
            return this.options.dashArray;
        },
        set: function (val) {
            this.setStyle({ dashArray: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "dashOffset", {
        get: function () {
            return this.options.dashOffset;
        },
        set: function (val) {
            this.setStyle({ dashOffset: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "fill", {
        get: function () {
            return this.options.fill;
        },
        set: function (val) {
            this.setStyle({ fill: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "fillColor", {
        get: function () {
            return this.options.fillColor;
        },
        set: function (val) {
            this.setStyle({ fillColor: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "fillOpacity", {
        get: function () {
            return this.options.fillOpacity;
        },
        set: function (val) {
            this.setStyle({ fillOpacity: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "fillRule", {
        get: function () {
            return this.options.fillRule;
        },
        set: function (val) {
            this.setStyle({ fillRule: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "className", {
        get: function () {
            return this.options.className;
        },
        set: function (val) {
            this.setStyle({ className: val });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "style", {
        get: function () {
            return this.options;
        },
        set: function (val) {
            this.setStyle(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "display", {
        get: function () {
            var container;
            try {
                container = this.getElement();
            }
            catch (err) {
                /* istanbul ignore next */
                return false;
            }
            return container.style.display !== 'none' && !!container.parentElement;
        },
        set: function (val) {
            var isDisplayed = this.display;
            if (isDisplayed === val) {
                return;
            }
            var container;
            try {
                container = this.getElement();
            }
            catch (err) {
                /* istanbul ignore next */
                return;
            }
            this.displayChange.emit(val);
            container.style.display = val ? '' : 'none';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "interactive", {
        get: function () {
            return this.options.interactive;
        },
        set: function (val) {
            var map = this._map;
            this.options.interactive = val;
            this.onRemove(map);
            this.onAdd(map);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleMarkerDirective.prototype, "properties", {
        get: function () {
            return this.feature.properties;
        },
        set: function (val) {
            this.feature.properties = val;
            this.geoJSONChange.emit(this.geoJSON);
        },
        enumerable: true,
        configurable: true
    });
    return CircleMarkerDirective;
}(leaflet_1.CircleMarker));
CircleMarkerDirective.decorators = [
    { type: core_1.Directive, args: [{
                selector: 'yaga-circle-marker',
            },] },
];
/** @nocollapse */
CircleMarkerDirective.ctorParameters = function () { return [
    { type: map_component_1.MapComponent, decorators: [{ type: core_1.Inject, args: [core_1.forwardRef(function () { return map_component_1.MapComponent; }),] },] },
]; };
CircleMarkerDirective.propDecorators = {
    'displayChange': [{ type: core_1.Output },],
    'strokeChange': [{ type: core_1.Output },],
    'colorChange': [{ type: core_1.Output },],
    'weightChange': [{ type: core_1.Output },],
    'opacityChange': [{ type: core_1.Output },],
    'lineCapChange': [{ type: core_1.Output },],
    'lineJoinChange': [{ type: core_1.Output },],
    'dashArrayChange': [{ type: core_1.Output },],
    'dashOffsetChange': [{ type: core_1.Output },],
    'fillChange': [{ type: core_1.Output },],
    'fillColorChange': [{ type: core_1.Output },],
    'fillOpacityChange': [{ type: core_1.Output },],
    'fillRuleChange': [{ type: core_1.Output },],
    'classNameChange': [{ type: core_1.Output },],
    'styleChange': [{ type: core_1.Output },],
    'positionChange': [{ type: core_1.Output },],
    'latChange': [{ type: core_1.Output },],
    'lngChange': [{ type: core_1.Output },],
    'radiusChange': [{ type: core_1.Output },],
    'geoJSONChange': [{ type: core_1.Output },],
    'addEvent': [{ type: core_1.Output, args: ['add',] },],
    'removeEvent': [{ type: core_1.Output, args: ['remove',] },],
    'popupopenEvent': [{ type: core_1.Output, args: ['popupopen',] },],
    'popupcloseEvent': [{ type: core_1.Output, args: ['popupclose',] },],
    'tooltipopenEvent': [{ type: core_1.Output, args: ['tooltipopen',] },],
    'tooltipcloseEvent': [{ type: core_1.Output, args: ['tooltipclose',] },],
    'clickEvent': [{ type: core_1.Output, args: ['click',] },],
    'dbclickEvent': [{ type: core_1.Output, args: ['dbclick',] },],
    'mousedownEvent': [{ type: core_1.Output, args: ['mousedown',] },],
    'mouseoverEvent': [{ type: core_1.Output, args: ['mouseover',] },],
    'mouseoutEvent': [{ type: core_1.Output, args: ['mouseout',] },],
    'contextmenuEvent': [{ type: core_1.Output, args: ['contextmenu',] },],
    'popupDirective': [{ type: core_1.Optional }, { type: core_1.ContentChild, args: [popup_directive_1.PopupDirective,] },],
    'tooltipDirective': [{ type: core_1.Optional }, { type: core_1.ContentChild, args: [tooltip_directive_1.TooltipDirective,] },],
    'position': [{ type: core_1.Input },],
    'lat': [{ type: core_1.Input },],
    'lng': [{ type: core_1.Input },],
    'radius': [{ type: core_1.Input },],
    'geoJSON': [{ type: core_1.Input },],
    'opacity': [{ type: core_1.Input },],
    'stroke': [{ type: core_1.Input },],
    'color': [{ type: core_1.Input },],
    'weight': [{ type: core_1.Input },],
    'lineCap': [{ type: core_1.Input },],
    'lineJoin': [{ type: core_1.Input },],
    'dashArray': [{ type: core_1.Input },],
    'dashOffset': [{ type: core_1.Input },],
    'fill': [{ type: core_1.Input },],
    'fillColor': [{ type: core_1.Input },],
    'fillOpacity': [{ type: core_1.Input },],
    'fillRule': [{ type: core_1.Input },],
    'className': [{ type: core_1.Input },],
    'style': [{ type: core_1.Input },],
    'display': [{ type: core_1.Input },],
    'interactive': [{ type: core_1.Input },],
    'properties': [{ type: core_1.Input },],
};
exports.CircleMarkerDirective = CircleMarkerDirective;
